[WebXR] Implement 'webgpu' on XRSessionInit

mwyrzykowski opened the pull request:
#### 5865de4721b26e5bb49242029946d8ba60559b44
<pre>
[WebXR] Implement &apos;webgpu&apos; on XRSessionInit
<a href="https://bugs.webkit.org/show_bug.cgi?id=296342">https://bugs.webkit.org/show_bug.cgi?id=296342</a>
<a href="https://rdar.apple.com/156435621">rdar://156435621</a>

Reviewed by Dan Glastonbury.

This gets <a href="https://immersive-web.github.io/webxr-samples/webgpu/vr-barebones.html">https://immersive-web.github.io/webxr-samples/webgpu/vr-barebones.html</a> functioning
from the W3C site.

Expand on some error messages as a bit of debugging effort was due to opaque
error messages.

Switch from bgra-srgb to bgra default format as WebGPU doesn&apos;t support configure with
bgra, rather the suggested approach is to pass the bgra-srgb format as a compatible view
format.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRBindingImpl.cpp:
(WebCore::WebGPU::XRBindingImpl::getPreferredColorFormat):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRSubImageImpl.cpp:
(WebCore::WebGPU::fromBacking):
(WebCore::WebGPU::XRSubImageImpl::colorTexture):
(WebCore::WebGPU::XRSubImageImpl::depthStencilTexture):
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
(WebCore::WebXRSystem::isFeaturePermitted const):
* Source/WebCore/Modules/webxr/XRGPUBinding.cpp:
(WebCore::XRGPUBinding::getViewSubImage):
(WebCore::XRGPUBinding::getPreferredColorFormat):
* Source/WebCore/Modules/webxr/XRGPUProjectionLayerInit.h:
* Source/WebCore/Modules/webxr/XRGPUSubImage.cpp:
(WebCore::XRGPUSubImage::XRGPUSubImage):
(WebCore::m_depthFormat):
(WebCore::XRGPUSubImage::colorTexture):
* Source/WebCore/Modules/webxr/XRGPUSubImage.h:
(WebCore::XRGPUSubImage::create):
* Source/WebCore/Modules/webxr/XRReferenceSpaceType.idl:
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::sessionFeatureFromReferenceSpaceType):
(PlatformXR::parseSessionFeatureDescriptor):
(PlatformXR::sessionFeatureDescriptor):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::addTextureToActiveResources):
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::getXRViewSubImage):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::addTextureToActiveResources):
(WebGPU::RenderPassEncoder::errorValidatingColorDepthStencilTargets const):
(WebGPU::RenderPassEncoder::errorValidatingPipeline const):
(WebGPU::RenderPassEncoder::colorDepthStencilTargetsMatch const): Deleted.
* Source/WebGPU/WebGPU/RenderPipeline.h:
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::errorValidatingColorDepthStencilTargets const):
(WebGPU::RenderPipeline::colorDepthStencilTargetsMatch const): Deleted.
* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::textureFormat):
* Source/WebGPU/WebGPU/XRProjectionLayer.h:
(WebGPU::XRProjectionLayer::create):
(WebGPU::XRProjectionLayer::colorFormat const):
(WebGPU::XRProjectionLayer::optionalDepthStencilFormat const):
(WebGPU::XRProjectionLayer::flags const):
(WebGPU::XRProjectionLayer::scale const):
* Source/WebGPU/WebGPU/XRProjectionLayer.mm:
(WebGPU::XRProjectionLayer::XRProjectionLayer):
(WebGPU::XRBinding::createXRProjectionLayer):
* Source/WebGPU/WebGPU/XRSubImage.h:
* Source/WebGPU/WebGPU/XRSubImage.mm:
(WebGPU::XRSubImage::update):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::XRDeviceProxy):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::toWKXRSessionFeatureFlags):
(WebKit::UIDelegate::UIClient::supportedXRSessionFeatures):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteXRBindingProxy.cpp:
(WebKit::WebGPU::RemoteXRBindingProxy::getPreferredColorFormat):

Canonical link: <a href="https://commits.webkit.org/298031@main">https://commits.webkit.org/298031@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9f1e549287a58f0f9c26fbfe7cbc47b97a7e8a72

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113918 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24129 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120083 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64698 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115807 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34298 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42231 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86578 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41620 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116866 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27298 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102317 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66955 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26496 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20447 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63796 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96666 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20564 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123313 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40958 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30499 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95413 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41331 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98529 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95262 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24283 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40324 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18134 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37094 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40831 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46337 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40465 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43765 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42222 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/Modules/webxr/XRGPUSubImage.cpp:
```diff
@@ -82,7 +81,7 @@ ExceptionOr<Ref<GPUTexture>> XRGPUSubImage::colorTexture()
         return Exception { ExceptionCode::InvalidStateError };
 
-    return GPUTexture::create(texture.releaseNonNull(), textureDescriptor(GPUTextureFormat::Bgra8unormSRGB), m_device);
+    return GPUTexture::create(texture.releaseNonNull(), textureDescriptor(GPUTextureFormat::Bgra8unorm, m_width, m_height), m_device);
```
This is `Bgra8unorm` value is hard-coded in a lot of places. Should it be queried from some configuration state instead?

mwyrzykowski replied:
Yes, that is correct. I plan to do so when adding HDR support https://bugs.webkit.org/show_bug.cgi?id=296675

There is also the issue that bgra8-srgb isn't a valid preferred canvas format in WebGPU but I think that is the default on visionPro. Rather, bgra8 is returned to JS and the requester asks for a canvas format of bgra8 with a view format of bgra8-srgb. I need to think how to make that work correctly as well

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebXR] Implement 'webgpu' on XRSessionInit
https://bugs.webkit.org/show_bug.cgi?id=296342
rdar://156435621

Reviewed by Dan Glastonbury.

This gets https://immersive-web.github.io/webxr-samples/webgpu/vr-barebones.html functioning
from the W3C site.

Expand on some error messages as a bit of debugging effort was due to opaque
error messages.

Switch from bgra-srgb to bgra default format as WebGPU doesn't support configure with
bgra, rather the suggested approach is to pass the bgra-srgb format as a compatible view
format.

* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRBindingImpl.cpp:
(WebCore::WebGPU::XRBindingImpl::getPreferredColorFormat):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUXRSubImageImpl.cpp:
(WebCore::WebGPU::fromBacking):
(WebCore::WebGPU::XRSubImageImpl::colorTexture):
(WebCore::WebGPU::XRSubImageImpl::depthStencilTexture):
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
(WebCore::WebXRSystem::isFeaturePermitted const):
* Source/WebCore/Modules/webxr/XRGPUBinding.cpp:
(WebCore::XRGPUBinding::getViewSubImage):
(WebCore::XRGPUBinding::getPreferredColorFormat):
* Source/WebCore/Modules/webxr/XRGPUProjectionLayerInit.h:
* Source/WebCore/Modules/webxr/XRGPUSubImage.cpp:
(WebCore::XRGPUSubImage::XRGPUSubImage):
(WebCore::m_depthFormat):
(WebCore::XRGPUSubImage::colorTexture):
* Source/WebCore/Modules/webxr/XRGPUSubImage.h:
(WebCore::XRGPUSubImage::create):
* Source/WebCore/Modules/webxr/XRReferenceSpaceType.idl:
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::sessionFeatureFromReferenceSpaceType):
(PlatformXR::parseSessionFeatureDescriptor):
(PlatformXR::sessionFeatureDescriptor):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::addTextureToActiveResources):
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::getXRViewSubImage):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::addTextureToActiveResources):
(WebGPU::RenderPassEncoder::errorValidatingColorDepthStencilTargets const):
(WebGPU::RenderPassEncoder::errorValidatingPipeline const):
(WebGPU::RenderPassEncoder::colorDepthStencilTargetsMatch const): Deleted.
* Source/WebGPU/WebGPU/RenderPipeline.h:
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::errorValidatingColorDepthStencilTargets const):
(WebGPU::RenderPipeline::colorDepthStencilTargetsMatch const): Deleted.
* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Texture::textureFormat):
* Source/WebGPU/WebGPU/XRProjectionLayer.h:
(WebGPU::XRProjectionLayer::create):
(WebGPU::XRProjectionLayer::colorFormat const):
(WebGPU::XRProjectionLayer::optionalDepthStencilFormat const):
(WebGPU::XRProjectionLayer::flags const):
(WebGPU::XRProjectionLayer::scale const):
* Source/WebGPU/WebGPU/XRProjectionLayer.mm:
(WebGPU::XRProjectionLayer::XRProjectionLayer):
(WebGPU::XRBinding::createXRProjectionLayer):
* Source/WebGPU/WebGPU/XRSubImage.h:
* Source/WebGPU/WebGPU/XRSubImage.mm:
(WebGPU::XRSubImage::update):
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::XRDeviceProxy):
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::toWKXRSessionFeatureFlags):
(WebKit::UIDelegate::UIClient::supportedXRSessionFeatures):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteXRBindingProxy.cpp:
(WebKit::WebGPU::RemoteXRBindingProxy::getPreferredColorFormat):

Canonical link: https://commits.webkit.org/298031@main
